zio-http/src/main/scala/zio/http/Route.scala

```diff
@@ -0,0 +1,486 @@
+  ): Handler[R1, Err1, In1, Out1] =
+    Handler
+      .fromFunctionZIO[In1] { in =>
+        self.run(in).toZIO.map { handler =>
+          if (handler ne null) handler
+          else default
+        }
+      }
+      .flatten
+
+  def toHandlerOrNull(
+    in: In,
+  )(implicit unsafe: Unsafe): HExit[R, Err, Handler[R, Err, In, Out]] = // NOTE: Handler[R, Err, In, Out] can be null
+    run(in)
+
+  final def toHandler(in: In): HExit[R, Err, Option[Handler[R, Err, In, Out]]] =
+    self.run(in).map(Option(_))
+
+  final def toHExitOrNull(in: In)(implicit unsafe: Unsafe): HExit[R, Err, Out] = { // NOTE: Out can be null
+    self.toHandlerOrNull(in).flatMap { handler =>
+      if (handler ne null) handler.apply(in)
+      else HExit.succeed(null).asInstanceOf[HExit[R, Err, Out]]
+    }
+  }
+
+  final def toSocketApp(implicit
+    ev1: WebSocketChannelEvent <:< In,
+    ev2: Err <:< Throwable,
+    trace: Trace,
+  ): SocketApp[R] =
+    SocketApp(event =>
+      self.toZIO(event).catchAll {
+        case Some(value) => ZIO.fail(value)
+        case None        => ZIO.unit
+      },
+    )
+
+  final def toZIO(in: In)(implicit trace: Trace): ZIO[R, Option[Err], Out] =
+    toHExitOrNull(in)(Unsafe.unsafe)
+      .mapError(Some(_))
+      .flatMap { out =>
+        if (out != null) HExit.succeed(out)
+        else HExit.fail(None)
+      }
+      .toZIO
+
+  final def withMiddleware[R1 <: R, In1 <: In, In2, Out2](
+    middleware: api.Middleware[R1, In2, Out2],
+  )(implicit ev1: In1 <:< Request, ev2: Out <:< Response): HttpRoute[R1, Err] =
+    middleware(self.asInstanceOf[HttpRoute[R, Err]])
+
+  final def when[In1 <: In](f: In1 => Boolean)(implicit trace: Trace): Route[R, Err, In1, Out] =
+    Route.fromHandlerHExit[In1] { in =>
+      if (f(in)) self.run(in)
+      else HExit.succeed(null)
+    }
+
+  final def whenZIO[R1 <: R, Err1 >: Err, In1 <: In](
+    f: In1 => ZIO[R1, Err1, Boolean],
+  )(implicit trace: Trace): Route[R1, Err1, In1, Out] =
+    Route.fromHandlerZIO { (in: In1) =>
+      f(in).mapError(Some(_)).flatMap {
+        case true  =>
+          self.run(in).toZIO.mapError(Some(_)).flatMap { handler =>
+            if (handler eq null) ZIO.fail(None)
+            else ZIO.succeed(handler)
+          }
+        case false =>
+          ZIO.fail(None)
+      }
+    }
+
+  final def withDefaultErrorResponse(implicit trace: Trace, ev1: Request <:< In, ev2: Out <:< Response): App[R] =
+    self.mapError { _ =>
+      Response(status = Status.InternalServerError)
+    }.asInstanceOf[App[R]]
+}
+
+object Route {
+
+  def collect[In]: Collect[In] = new Collect[In](())
+
+  def collectHandler[In]: CollectHandler[In] = new CollectHandler[In](())
+
+  def collectHExit[In]: CollectHExit[In] = new CollectHExit[In](())
+
+  def collectRoute[In]: CollectRoute[In] = new CollectRoute[In](())
+
+  def collectZIO[In]: CollectZIO[In] = new CollectZIO[In](())
+
+  def empty: Route[Any, Nothing, Any, Nothing] =
+    (_: Any) => HExit.succeed(null)
```

Comment:
`HExit.succeed(null)`?

Is this truly an empty with respect to `++`?

Reply:
Yes as I mentioned somewhere in the long PR description I removed `HExit.empty` (so users can't use it to define `Handler`s but did not introduce a `HExitOrEmpty` or used `Option` but in `Route` `HExit.succeed(null)` represents the unhandled case.

Maybe it's an unnecessary premature optimization, I would be happy to not do that way.